app/Models/Chat/UserChannel.php

```diff
@@ -89,91 +84,105 @@ public static function presenceForUser(User $user)
+            // this says != PUBLIC but really is just == PM because of the data loaded.
+            if ($channel->type !== Channel::TYPES['public']) {
+                $userIds = $userIdsByChannelId[$channel->getKey()] ?? [];
+
+                foreach ($userIds as $userId) {
+                    if ($usersById[$userId] ?? null) {
+                        $channelUserIds[] = $userId;
+                    }
                 }
+            }
 
-                if ($userChannel->type === Channel::TYPES['pm']) {
-                    // remove ourselves from $membersArray, leaving only the other party
-                    $members = array_diff($filteredChannelMembers, [$userId]);
-                    $targetUser = $byUserId[array_shift($members)] ?? null;
+            $presence['users'] = $channelUserIds;
 
-                    // hide if target is restricted ($targetUser missing) or is blocked ($targetUser->foe)
-                    if (!$targetUser || $targetUser->foe) {
-                        return [];
-                    }
+            if ($channel->type === Channel::TYPES['pm']) {
+                // remove ourselves from $channelUserIds, leaving only the other party.
+                // array_shift doesn't require array_values to be called first.
+                $members = array_diff($channelUserIds, [$user->getKey()]);
+                $targetUser = $usersById[array_shift($members)] ?? null;
 
-                    // override channel icon and display name in PMs to always show the other party
-                    $userActual = $targetUser->userScoped;
-                    $presence['icon'] = $userActual->user_avatar;
-                    $presence['name'] = $userActual->username;
-                    // ideally this should be ChatChannelSend but it involves too many queries
-                    $presence['moderated'] = $presence['moderated'] || !priv_check_user($user, 'ChatStart', $userActual)->can();
+                // hide if target is restricted or blocked unless blocked user is a moderator.
+                if (!$targetUser
+                    || $user->hasBlocked($targetUser) && !$targetUser->isModerator()) {
+                    return [];
                 }
 
-                return $presence;
+                // override channel icon and display name in PMs to always show the other party
+                $presence['icon'] = $targetUser->user_avatar;
+                $presence['name'] = $targetUser->username;
+                // ideally this should be ChatChannelSend but it involves too many queries
+                $presence['moderated'] = $presence['moderated'] || !priv_check_user($user, 'ChatStart', $targetUser)->can();
             }
-        );
+
+            return $presence;
+        });
 
         // strip out the empty [] elements (from restricted/blocked users)
         return array_values(array_filter($collection));
     }
+
+    private static function userChannelsForUser(User $user)
```

Comment:
`forUser` is enough, isn't it? UserChannel is already its class name after all. And maybe as scope?